### PR TITLE
Use sphinx.ext.linkcode for more precise source code links

### PR DESCRIPTION
**Pull request by melechlapson**
### Summary
This addresses https://github.com/Qiskit/documentation/issues/517 by switching out the sphinx.ext.viewcode for the sphinx.ext.linkcode extension which allows our GitHub links in the documentation to be linked to the specific lines of code, not just the file. This was tested successfully in https://github.com/Qiskit/qiskit_sphinx_theme/pull/589 and in the qiskit repo in https://github.com/Qiskit/qiskit/pull/11851 and now we want to implement it in the runtime repo.

The API generation script at Qiskit/documentation already knows how to handle sphinx.ext.linkcode.

Example links from this PR:

https://github.com/Qiskit/qiskit-ibm-runtime/tree/main/qiskit_ibm_runtime/fake_provider/backends/armonk/fake_armonk.py#L21-L33

https://github.com/Qiskit/qiskit-ibm-runtime/tree/main/qiskit_ibm_runtime/fake_provider/fake_backend.py#L269-L282

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -117,3 +119,85 @@
...
+        "GITHUB_REF_NAME" not in os.environ
+        and "BUILD_SOURCE_BRANCH_NAME" not in os.environ
+    ):
+        return "main"
+
+    # PR workflows set the branch they're merging into.
+    if base_ref := (
+        os.environ.get("GITHUB_BASE_REF")
+        or os.environ.get("SYSTEM_PULL_REQUEST_TARGET_BRANCH_NAME")
+    ):
```
Comment (Eric-Arellano) [inline docs/conf.py:146]: Bad leftover code from Azure

```suggestion
    if base_ref := os.environ.get("GITHUB_BASE_REF")
```

File: docs/conf.py [hunk 2]
```diff
@@ -117,3 +119,85 @@
...
+        return base_ref
+
+    ref_name = (
+        os.environ.get("GITHUB_REF_NAME")
+        or os.environ.get("BUILD_SOURCE_BRANCH_NAME")
+    )
+    assert ref_name is not None
```
Comment (Eric-Arellano) [inline docs/conf.py:153]:
```suggestion
    ref_name = os.environ["GITHUB_REF_NAME"]
```

File: docs/conf.py [hunk 3]
```diff
@@ -117,3 +119,85 @@
...
+
+    # Check if the ref_name is a tag like `1.0.0` or `1.0.0rc1`. If so, we need
+    # to transform it to a Git branch like `stable/1.0`.
+    version_without_patch_match = re.match(r"(\d+\.\d+)", ref_name)
+    return (
+        f"stable/{version_without_patch_match.group()}"
+        if version_without_patch_match
+        else ref_name
+    )
```
Comment (Eric-Arellano) [inline docs/conf.py:162]:
```suggestion
    version_without_patch = re.match(r"(\d+\.\d+)", ref_name)
    return (
        f"stable/{version_without_patch.group()}"
        if version_without_patch
        else ref_name
    )
```

File: docs/conf.py [hunk 4]
```diff
@@ -117,3 +119,85 @@
...
+GITHUB_BRANCH = determine_github_branch()
+
+
+def linkcode_resolve(domain, info):
+    if domain != "py":
+        return None
+
+    module_name = info["module"]
+    module = sys.modules.get(module_name)
+    if module is None or "qiskit" not in module_name:
```
Comment (Eric-Arellano) [inline docs/conf.py:174]:
```suggestion
    if module is None or "qiskit_ibm_runtime" not in module_name:
```

File: docs/conf.py [hunk 5]
```diff
@@ -117,3 +119,85 @@
...
+        )
+        if not is_valid_code_object:
+            return None
+    try:
+        full_file_name = inspect.getsourcefile(obj)
+    except TypeError:
+        return None
+    if full_file_name is None:
+        return None
+    file_name = full_file_name.split("/qiskit-ibm-runtime/")[-1]
```
Comment (Eric-Arellano) [inline docs/conf.py:194]: The files belong to this folder: https://github.com/Qiskit/qiskit-ibm-runtime/tree/main/qiskit_ibm_runtime

```suggestion
    file_name = full_file_name.split("/qiskit_ibm_runtime/")[-1]
```

File: docs/conf.py [hunk 6]
```diff
@@ -117,3 +119,85 @@
...
+
+    try:
+        source, lineno = inspect.getsourcelines(obj)
+    except (OSError, TypeError):
+        linespec = ""
+    else:
+        ending_lineno = lineno + len(source) - 1
+        linespec = f"#L{lineno}-L{ending_lineno}"
+    return f"https://github.com/Qiskit/qiskit-ibm-runtime/tree/{GITHUB_BRANCH}/qiskit-ibm-runtime/{file_name}{linespec}"
```
Comment (Eric-Arellano) [inline docs/conf.py:203]:
```suggestion
    return f"https://github.com/Qiskit/qiskit-ibm-runtime/tree/{GITHUB_BRANCH}/qiskit_ibm_runtime/{file_name}{linespec}"
```

**Review by Eric-Arellano — COMMENTED** [review]
Thanks! You need to update `passenv` in `tox.ini`.

File: docs/conf.py [hunk 7]
```diff
@@ -117,3 +119,85 @@
...
+
+    We need to decide whether to use `stable/<version>` vs. `main` for dev builds.
+    Refer to https://docs.github.com/en/actions/learn-github-actions/variables
+    for how we determine this with GitHub Actions.
+    """
+    # If CI env vars not set, default to `main`. This is relevant for local builds.
+    if (
+        "GITHUB_REF_NAME" not in os.environ
+        and "BUILD_SOURCE_BRANCH_NAME" not in os.environ
+    ):
```
Comment (Eric-Arellano) [inline docs/conf.py:139]:
```suggestion
    if "GITHUB_REF_NAME" not in os.environ:
```

**Review by Eric-Arellano — APPROVED** [review]
Thanks! The CI artifact at https://github.com/Qiskit/qiskit-ibm-runtime/actions/runs/8069238711/artifacts/1279793894 looks great.

File: docs/conf.py [hunk 8]
```diff
@@ -117,3 +119,75 @@
...
+        is_valid_code_object = (
+            inspect.isclass(obj) or inspect.ismethod(obj) or inspect.isfunction(obj)
+        )
+        if not is_valid_code_object:
+            return None
+    try:
+        full_file_name = inspect.getsourcefile(obj)
+    except TypeError:
+        return None
+    if full_file_name is None:
```
Comment (Eric-Arellano) [inline docs/conf.py:182]:
```suggestion
    if full_file_name is None or `/qiskit_ibm_runtime/` not in full_file_name:
```

**Review by frankharkins — APPROVED** [review]
Looks good, thanks!

**Pull request merged**